[WIP] Notebook comparing FSS and secureNN as currently implemented SMPC protocols. Solving #4063

NiWaRe opened the pull request:
## Description
This notebook will give an overview and a very quick explanation over/of the different SMPC protocols that are currently implemented in PySyft. It'll also elaborate on **what kind of Machine Learning you can conduct in an encrypted fashion using them**, along a comparison of their resulting performance to each other and to the non-encrypted scenario. The protocol explanations should mainly serve to give a high-level understanding of what the main **crypto-slang** stands for, how the terms relate to each other and what ressources serve as good starting points to dig deeper! 

Closes #4063

* **Current status:** After familiarizing with the different protocols on a high level I've written the explanations which aim at giving a quick overview of the protocols and should give a quick understanding of what different models, optims, losses can be used with the different protocols. 
* **Next steps:** Incorporate feedback from @LaRiffle @gmuraru @somiljain7 if any. Find suitable examples and implement them using PySyft (incl. additional explanations). Use these + explanations in papers to explain/showcase the performance of the protocols for different computations/applications. 

## Affected Dependencies
None.

## How has this been tested?
- Follow instructions in notebook and run cells. 

## Checklist 
*Still WIP.*
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

Nilanshrajput commented:
@NiWaRe  this is awesome!!!

gmuraru commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>TYPO: the first category -&gt; the first category.</p><p>At point 3, I think in our scenario we do the reconstruct only on a "worker" (but all workers can do that reconstruct)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

NiWaRe replied:
<p>Alright typos fixed, I'll remember to paste everything into GDocs from now on :D </p><p>I've also added the two remarks about that only one worker does the constructing in our case and about the additional communication step. </p>

gmuraru commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>TYPO: randomely -&gt; randomly</p><p>Maybe we can also add that the offline phase exists to speed up the online phase (you actually pointed this out).</p><p>But, we can also generate Beaver Triples during the online phase - but this would impact the <code>online</code>  performance</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

NiWaRe replied:
<p>Ok thanks, I'll make that clearer!</p>

gmuraru commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>NITPICK: Sounds a little off "the server that serves" - maybe "the party that serves"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

NiWaRe replied:
<p>:D Aight!</p>

gmuraru commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>TYPO: as a circuit consisting of different logical gates (e.g. AND, XOR, etc.). &lt;-- dot at the end :D</p><p>withouut -&gt; without</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

gmuraru commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>Q: I think it might be better to provide the link from arxiv</p><p>TYPO: a dot after (state-of-the-art ML protocol before SecureNN).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

NiWaRe replied:
<p>I didn't find the paper on arxiv, but I changed the link to the link mentioned in the code. </p>

gmuraru replied:
<p>Ahh - sorry. It was here: https://eprint.iacr.org/2017/396.pdf</p>

gmuraru commented:
Took a quick look over it and left some comments! Need to read more and will come back with more feedback :D

omershlo commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p><em> circuit-gabling-based SMPC -&gt;  circuit-garbling-based SMPC</em></p><p><em>consider adding a link to: </em></p><p><em>GC wiki: </em>https://wiki.mpcalliance.org/garbled_circuit.html </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

NiWaRe replied:
<p>Thanks, very helpful website!</p>

omershlo commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>1) typo in efficent</p><p> </p><p>2) If the protocol is for : 3 parties with one dishonest party in the malicious adversary model - how is this protocol be made to work with 2 parties + 1 crypto-worker? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

NiWaRe replied:
<p>Well the secureNN requires an honest majority for all parties<em> that store a share</em> - that doesn't include the crypto-worker. So 2 parties + 1 crypto-worker where one party is dishonest isn't considered a honest majority which means secureNN can't safely be used as far as I understand. </p>

LaRiffle replied:
Oh, interesting! Thanks for the explanation :)

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>"These are being trusted of not collaborating with each other" not necessarily! The exact term is colluding.</p><p>You can say that it is resistant to some level of collusion (we often consider that we have a <em>honest majority</em>), and can be<u> seen as</u> a <strong>trusted execution environment</strong> as long as thee isn't too much collusion. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

NiWaRe replied:
<p>Ok interesting, thanks! I'll change that.</p>

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>good explanation!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

NiWaRe replied:
<p>Thanks! </p>

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>there is also a version with 3 parties + 1 crypto provider I think in SecureNN</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

NiWaRe replied:
<p>Oh yes I made a mistake, in the paper they actually do consider 3 parties + 1 crypto provider. </p>

NiWaRe replied:
<p>They describe the system as a 3PC-protocol, and the algorithms always consider two parties holding the secret shares (referred to as P0, and P1) and one party generating the random primitives (P2).</p><p> </p><p>Do you have some resource pointing to the 4PC-version which I should include or should I simply state it? </p><p> </p>

LaRiffle replied:
Maybe you can just mention that the 4PC version exist without detailing yeah, maybe that's not the most important par of SecureNN (and we don't implement it!)

NiWaRe replied:
Sounds good, I'll do that!

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>For mat mul, it could be good to add that the beaver triples extend to matrices</p><p> </p><p>the "Low Level" components of SecureNN are not necessarily useful unless you explain precisely how they use each other to implement ReLU for example</p><p> </p><p>You mention that division is possible, it is indeed with SecureNN but it is incredibly slow -  you can mention it</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

NiWaRe replied:
<p>Alright, thanks! </p><p>Do you know why specifically the division is slow? </p><p> </p><p>Also the idea about the "Low Level" components was to give more advanced users the possibility to themselves assess whether special computations are possible with the implemented low level algorithms of the protocol. Detailed explanations about how the algorithms work might scare of the less experienced users, how about I could include the graph showing which low level components work together and are based on each other (Fig.3 in the paper). What do you think? </p>

LaRiffle replied:
<p>&gt; Do you know why specifically the division is slow?</p><p> </p><p>Yeah because it you check the implem, you can see that it makes roughly n call to the comparison, and because the comparison is already a bit expensive, it gets terrible^^</p><p> </p><p>Yeah Fig 3 is helpful I agree! But even for advanced users, we don't expect them to reuse parts of the low level components to build new computations, they should instead rely on the the Relu, Div, and MaxPool I would say</p>

NiWaRe replied:
Ahh ok because the division is implemented as subtraction and then bitwise comparison and shifting, right?

Ok so I think I'll just shift the low-level parts with the figures to an extra section at the end for users who are more interested and want to have a starting point to understand the implementations and the papers. 👍

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,390 @@
+{
```
<p>Might be interesting to point out that indeed FSS doesn't require lots of communication rounds. But is computationally more intensive than SecureNN (because we call a PRG <code>G</code> many times)</p><p> </p><p>For Dopout, i guess we could implement it, just no one took the time to do so :) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

LaRiffle replied:
<p>Erratum: we support dropout for secureNN + FSS</p>

NiWaRe replied:
<p>Alright perfect, I'll add that. </p>

NiWaRe commented:
Hey @LaRiffle @gmuraru @omershlo @Nilanshrajput, 
I incorporated your feedback and added the code, the first version should be completed now. 
For now, I've decided to compare the training of different simple models on the Cifar10 dataset and compare performance in terms of accuracy and speed on a very small test-set. (otherwise, it takes to long on my laptop)
I created it to also serve as a sort of sandbox to create try out different models, computations, loss functions, optimizers for different developers. One final thing I planned to add is the local training to have a certain benchmark against which to compare. 

Would be nice if you could give me some feedback on the approach, structure, comprehensibility, and correctness. Also, feel free to suggest or request some features you'd like to have in the notebook. 👍

LaRiffle commented on .gitignore:
```diff
@@ -21,4 +21,5 @@ docs/_modules
 .vscode/
 **/pip-wheel-metadata
 .ropeproject/
+./examples/tutorials/Part*
```
Don't add this line, the tutorials shouldn't be ignored :)

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1181 @@
+{
```
<p>similair -&gt; similar</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

LaRiffle replied:
<p>(Only inference, we use pre-trained models) -&gt; in the intro you mention doing training</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

LaRiffle replied:
<p>You can add a line to explain what all those are: (Wrapper)&gt;AutogradTensor&gt;FixedPrecisionTensor&gt;[AdditiveSharingTensor]  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

LaRiffle replied:
<p>you coud reuse your encryption kwargs in <code>secret_share</code>  </p><p> </p><p>CrossEntropyLoss is a function or  a class? Maybe add it to pysyft in a different PR :) </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

LaRiffle replied:
<p>"<em>This reduces communication complexity significantly." -&gt; can be removed</em></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

LaRiffle replied:
<blockquote><strong>Encryption with n parties possible (n &gt;= 2)</strong></blockquote><p>It mostly 2 party I'd say</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

LaRiffle replied:
<p>why is it commented?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1181 @@
+{
```
<blockquote>Does the compression also needs to be turned off, if so, why? (as in Théo's tuto)</blockquote><p>It makes the code faster, it's not compulsory</p><p> </p><blockquote>if resnet is in train() mode it can't run an encrypted forward pass</blockquote><p>Training resnet is not supported so far (and too expensive)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

NiWaRe replied:
<p>Alright, thanks for the explanations! </p>

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1181 @@
+{
```
<p>Do you really want to feature  remote encryption? Maybe just mention it as a side note as it makes just more code.</p><p>Also, if you want to use encryption_kwargs, use it on all cells also the previous one for example.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

NiWaRe replied:
<p>Concerning showcasing the remote encryption an alternative would also be that you include the local encryption part in your tutorial Part 11 bis - given of course you also think that this direct comparison would make sense and doesn't yet exist in a notebook. :+1</p>

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1181 @@
+{
```
<p>we don't see the loss reducing, why is so?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

NiWaRe replied:
<p>yes this the reason for the low acc in the last discussion you started I guess. Re-running the training, but still I don't understand why sometimes things like that occur, this is why I maybe suggested an issue with GarbageCollector in my notebook (because usually I don't get these kind of error if I restart the notebook and directly train only one model)</p>

NiWaRe replied:
<p>Ok I ran it again and the loss was 715 and 715 for both epochs, still the test-acc. was 0.14 which should be correct, so I guess the optim together with the quite small batch size just leads to bad first steps. </p>

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1181 @@
+{
```
<p>SSN + CELoss + Avg Pool is missing :(</p><p> </p><p>Test Accuracy is the same of SNN with MSE and CE ? I would expect the Accuracy of SNN + CE to match the one of FSS + CE</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

NiWaRe replied:
<blockquote>SSN + CELoss + Avg Pool is missing </blockquote><p>nearly there :) </p>

NiWaRe replied:
<blockquote>I would expect the Accuracy of SNN + CE to match the one of FSS + CE</blockquote><p>Re-training yielded 0.14 so all good. :+1</p>

LaRiffle commented on examples/tutorials/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1181 @@
+{
```
<p>You can remove the TODO: just the protocol calls a PRG many times :) </p><p> </p><p>Mention that this table is extracted from the AriaNN paper maybe</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

NiWaRe replied:
<p>Done. </p>

NiWaRe reviewed (commented):
Ok thanks for the feedback, I'll go over it and re-commit. +1

LaRiffle reviewed (commented):
Final comment: can you put the tutorial in /example/tutorial/advanced?
This would be awesome!

This is a really strong contribution

LaRiffle reviewed (approved):
🎉

gmuraru commented on examples/tutorials/advanced/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1112 @@
+{
```
<p>his means that usually it is considered that there exists a <em>honest majority</em> of servers which are trusted of not collaborating with each other</p><p>^^ should we also say here that they do not divert from what they should do in the protocol?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

NiWaRe replied:
<p>Will do that :+1</p>

gmuraru commented on examples/tutorials/advanced/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1112 @@
+{
```
<p>TYPO: I think here there is an extra "*"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

gmuraru replied:
<p>Q: Also for this iamge</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

gmuraru commented on examples/tutorials/advanced/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1112 @@
+{
```
<p>Add a dot (.) at the end of " about <a href="https://blog.openmined.org/federated-learning-types/" target="_blank">different kinds of Federated Learning</a>)"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

NiWaRe replied:
<p>Done.</p>

gmuraru commented on examples/tutorials/advanced/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1112 @@
+{
```
<p>(the first category) &lt;-- also a dot here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

NiWaRe replied:
<p>Done.</p>

gmuraru commented on examples/tutorials/advanced/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1112 @@
+{
```
<p>(i.e. only two servers are involved)  &lt;--- also here a dot at the end?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

NiWaRe replied:
<p>Also done :D </p>

gmuraru commented on examples/tutorials/advanced/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1112 @@
+{
```
<p>Q: I am using ReviewNB but the images are not appearing? (Should the link be updated or it would be ok after merge?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

NiWaRe replied:
<p>Yes forgot to update!</p>

gmuraru commented on examples/tutorials/advanced/SMPC_Protocols_Explanation.ipynb:
```diff
@@ -0,0 +1,1112 @@
+{
```
<p>Q: With the PySyft Projects should we better point here "https://github.com/OpenMined"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/4635/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

NiWaRe replied:
<p>Yes I agree, changed that. </p>

gmuraru commented:
You should rebase - there was a fix that got into the master branch

gmuraru reviewed (approved):
GG! Everything looks great!

Left only some nitpicking stuff.
